src/test/java/be/seeseemelk/mockbukkit/entity/PlayerMockTest.java

```diff
@@ -504,20 +508,162 @@ public void chat_AnyMessage_AsyncEventFired()
+	}
 
+	@Test
+	public void getTotalExperience_Default_EqualsZero()
+	{
+		assertEquals(0, player.getTotalExperience());
+	}
 
+	@Test
+	public void setLevel_SomeValue_LevelSetExactly()
+	{
+		player.setLevel(15);
+		assertEquals(15, player.getLevel());
+	}
 
+	@Test
+	public void setExp_SomeValue_LevelSetExactly()
+	{
+		player.setExp(0.5F);
+		assertEquals(0.5, player.getExp(), 0.5);
+	}
 
+	@Test(expected = IllegalArgumentException.class)
+	public void setExp_GreaterThanOne_ExceptionThrown()
+	{
+		player.setExp(1.1F);
+	}
 
+	@Test(expected = IllegalArgumentException.class)
+	public void setExp_LessThanZero_ExceptionThrown()
+	{
+		player.setExp(-1.0F);
+	}
 
+	@Test
+	public void setTotalExperience_SomeValue_TotalExpSetExactly()
+	{
+		player.setTotalExperience(100);
+		assertEquals(100, player.getTotalExperience());
+	}
 
+	@Test
+	public void setTotalExperience_NegativeValue_ClampedAtZero()
+	{
+		player.setTotalExperience(-200);
+		assertEquals(0, player.getTotalExperience(), 0);
+	}
 
+	@Test
+	public void getExpToLevel_CorrectExp()
+	{
+		for (int i = 0; i < expRequired.length; i++)
+		{
+			player.setLevel(i);
+			assertEquals(expRequired[i], player.getExpToLevel());
+		}
+	}
+
+	@Test
+	public void giveExpLevel_Negative_ClampedAtZero()
+	{
+		player.setExp(0.5F);
+		player.setLevel(1);
+		player.giveExpLevels(-100);
+		assertEquals(0, player.getExp(), 0);
+		assertEquals(0, player.getLevel());
+	}
+
+	@Test
+	public void giveExp_SomeExp_IncreaseLevel()
+	{
+		for (int i = 0; i < expRequired.length; i++)
+		{
+			assertEquals(0, player.getExp(), 0);
+			player.giveExp(expRequired[i]);
+			assertEquals(i+1, player.getLevel());
+		}
+	}
 
+	@Test
+	public void giveExp_SomeExp_IncreaseMultipleLevels()
+	{
+		player.giveExp(expRequired[0] + expRequired[1] + expRequired[2]);
+		assertEquals(3, player.getLevel());
+		assertEquals(expRequired[0] + expRequired[1] + expRequired[2], player.getTotalExperience(), 0);
+	}
 
+	@Test
+	public void giveExp_SomeExp_DecreaseLevel()
+	{
+		player.giveExp(expRequired[0] + expRequired[1]);
+		player.giveExp(-expRequired[1]);
+		assertEquals(1, player.getLevel());
+		assertEquals(expRequired[0], player.getTotalExperience(), 0);
+	}
 
+	@Test
+	public void giveExp_SomeExp_DecreaseMultipleLevels()
+	{
+		player.giveExp(expRequired[0] + expRequired[1]);
+		player.giveExp(-(expRequired[0] + expRequired[1]));
+		assertEquals(0, player.getLevel());
+		assertEquals(0.0, player.getTotalExperience(), 0);
+	}
 
+	@Test
+	public void giveExp_SomeLevelChange_LevelEventFired()
+	{
+		TestPlugin plugin = MockBukkit.load(TestPlugin.class);
+		AtomicInteger levelCount = new AtomicInteger();
+		Bukkit.getPluginManager().registerEvents(new Listener()
+		{
+			@EventHandler
+			public void onLevelChangeEvent(PlayerLevelChangeEvent event)
+			{
+				levelCount.incrementAndGet();
+			}
 
+			@EventHandler
+			public void onExpChangeEvent(PlayerExpChangeEvent event)
+			{
+				fail("PlayerExpChangeEvent should not be called");
```

Comment:
is this bukkit behavior? I looked at the javadocs and didn't find anything about that, as I user I would expect both events to be called (since the total exp changed?)

Reply:
According to [the docs](https://hub.spigotmc.org/javadocs/spigot/org/bukkit/event/player/PlayerExpChangeEvent.html), `PlayerExpChangeEvent` is only called "when a players experience changes naturally". Artifically calling `giveExp` through the API is not natural behavior. I also found [a thread](https://bukkit.org/threads/playerexpchangeevent-not-running.367378/) where somebody asked about this behavior themselves.